distributed/shuffle/_worker_extension.py

```diff
@@ -192,6 +145,318 @@ def heartbeat(self) -> dict[str, Any]:
+            rpc=rpc,
+            scheduler=scheduler,
+            memory_limiter_comms=memory_limiter_comms,
+            memory_limiter_disk=memory_limiter_disk,
+        )
+        self.old = old
+        self.new = new
+        partitions_of = defaultdict(list)
+        for part, addr in worker_for.items():
+            partitions_of[addr].append(part)
+        self.partitions_of = dict(partitions_of)
+        self.worker_for = worker_for
+        self._slicing = rechunk_slicing(old, new)
+        self._old_to_new = _old_to_new(old, new)
+
+    async def _receive(self, data: list[tuple[ArrayRechunkShardID, bytes]]) -> None:
+        self.raise_if_closed()
+
+        buffers = defaultdict(list)
+        for d in data:
+            id, payload = d
+            if id in self.received:
+                continue
+            self.received.add(id)
+            self.total_recvd += sizeof(d)
+
+            buffers[id.new_index].append(payload)
+
+        del data
+        if not buffers:
+            return
+        try:
+            await self._write_to_disk(buffers)
+        except Exception as e:
+            self._exception = e
+            raise
+
+    async def add_partition(self, data: np.ndarray, input_partition: NIndex) -> int:
+        self.raise_if_closed()
+        if self.transferred:
+            raise RuntimeError(f"Cannot add more partitions to shuffle {self}")
+
+        def _() -> dict[str, list[tuple[ArrayRechunkShardID, bytes]]]:
+            """Return a mapping of worker addresses to a list of tuples of shard IDs
+            and shard data.
+
+            As shard data, we serialize the data together with the sub-index of the
+            slice within the new chunk to assemble the new chunk from its pieces.
+            """
+            out: dict[str, list[tuple[ArrayRechunkShardID, bytes]]] = defaultdict(list)
+            for id, nslice in self._slicing[input_partition]:
+                out[self.worker_for[id.new_index]].append(
+                    (id, pickle.dumps((id.sub_index, data[nslice])))
+                )
+            return out
+
+        out = await self.offload(_)
+        await self._write_to_comm(out)
+        return self.run_id
+
+    async def get_output_partition(self, i: NIndex) -> np.ndarray:
+        self.raise_if_closed()
+        assert self.transferred, "`get_output_partition` called before barrier task"
+
+        assert self.worker_for[i] == self.local_address, (
+            f"Output partition {i} belongs on {self.worker_for[i]}, "
+            f"not {self.local_address}. "
+        )
+        # ^ NOTE: this check isn't necessary, just a nice validation to prevent incorrect
+        # data in the case something has gone very wrong
+
+        await self.flush_receive()
+        data = self._read_from_disk(i)
+        subdims = tuple(len(self._old_to_new[dim][ix]) for dim, ix in enumerate(i))
+        with self.time("cpu"):
+            arr = convert_chunk(data, subdims)
```

Comment:
```suggestion
        arr = await self.offload(convert_chunk, data, subdims)
```
pretty sure this should be offloaded. you even instrument it with CPU time so this is likely nontrivial

Reply:
Good catch, I just noticed we don't do that on the P2P shuffle either (which is probably why I haven't done so here).